libloki/api.js

```diff
@@ -1,47 +1,128 @@
+
+  // feel free to change this to window.console.warn to have a stack with all those logs
+  // disable loki logs
+  const debugLogFn = window.console.warn; // ;
+
+  function logSessionMessageSending(...args) {
+    if (debugFlags | DebugFlagsEnum.SESSION_MESSAGE_SENDING && debugLogFn) {
+      debugLogFn(...args);
+    }
+  }
+
+  function logGroupSync(...args) {
+    if (debugFlags | DebugFlagsEnum.GROUP_SYNC_MESSAGES && debugLogFn) {
+      debugLogFn(...args);
+    }
+  }
+
+  function logGroupRequestInfo(...args) {
+    if (debugFlags | DebugFlagsEnum.GROUP_REQUEST_INFO && debugLogFn) {
+      debugLogFn(...args);
+    }
+  }
+
+  function logContactSync(...args) {
+    if (debugFlags | DebugFlagsEnum.GROUP_CONTACT_MESSAGES && debugLogFn) {
+      debugLogFn(...args);
+    }
+  }
+
+  function logAutoFriendRequest(...args) {
+    if (
+      debugFlags | DebugFlagsEnum.AUTO_FRIEND_REQUEST_MESSAGES &&
+      debugLogFn
+    ) {
+      debugLogFn(...args);
+    }
+  }
+
+  function logNormalFriendRequest(...args) {
+    if (
+      debugFlags | DebugFlagsEnum.NORMAL_FRIEND_REQUEST_MESSAGES &&
+      debugLogFn
+    ) {
+      debugLogFn(...args);
+    }
+  }
+
+  function logSessionRequest(...args) {
+    if (debugFlags | DebugFlagsEnum.SESSION_REQUEST_MESSAGES && debugLogFn) {
+      debugLogFn(...args);
+    }
   }
 
-  async function sendOnlineBroadcastMessage(pubKey, isPing = false) {
+  function logBackgroundMessage(...args) {
+    if (debugFlags | DebugFlagsEnum.SESSION_BACKGROUND_MESSAGE && debugLogFn) {
+      debugLogFn(...args);
+    }
+  }
+
+  // Returns the primary device pubkey for this secondary device pubkey
+  // or the same pubkey if there is no other device
+  async function getPrimaryDevicePubkey(pubKey) {
     const authorisation = await window.libloki.storage.getGrantAuthorisationForSecondaryPubKey(
       pubKey
     );
-    if (authorisation && authorisation.primaryDevicePubKey !== pubKey) {
-      sendOnlineBroadcastMessage(authorisation.primaryDevicePubKey);
+    return authorisation ? authorisation.primaryDevicePubKey : pubKey;
+  }
+
+  async function sendBackgroundMessage(pubKey, debugMessageType) {
+    const primaryPubKey = await getPrimaryDevicePubkey(pubKey);
+    if (primaryPubKey !== pubKey) {
+      // if we got the secondary device pubkey first,
+      // call ourself again with the primary device pubkey
+      await sendBackgroundMessage(primaryPubKey, debugMessageType);
       return;
     }
-    const p2pAddress = null;
-    const p2pPort = null;
-    // We result loki address message for sending "background" messages
-    const type = textsecure.protobuf.LokiAddressMessage.Type.HOST_UNREACHABLE;
 
-    const lokiAddressMessage = new textsecure.protobuf.LokiAddressMessage({
-      p2pAddress,
-      p2pPort,
-      type,
-    });
-    const content = new textsecure.protobuf.Content({
-      lokiAddressMessage,
-    });
+    const backgroundMessage = textsecure.OutgoingMessage.buildBackgroundMessage(
```

Comment:
separate the build and the sending. so we can test one and another with different test cases (also data after sending is encrypted so not super useful)